pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,554 @@
+        _ = tokio::time::sleep_until(sleep_until) => ControlFlow::Continue(()),
+        _ = cancel.cancelled() => {
+            info!("shutting down");
+            ControlFlow::Break(())
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+enum IterationOutcome {
+    NoPressure,
+    Cancelled,
+    Finished(IterationOutcomeFinished),
+}
+
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+#[derive(Debug)]
+struct IterationOutcomeFinished {
+    before: UsageWithPressure,
+    after_actual: UsageWithPressure,
+    after_expected: Usage,
+
+    // Ok(false) a bit problematic
+    evictions_failed: LayerCount,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Default)]
+struct LayerCount {
+    file_sizes: u64,
+    count: usize,
+}
+
+async fn disk_usage_eviction_task_iteration_impl(
+    task_config: &DiskUsageEvictionTaskConfig,
+    storage: &GenericRemoteStorage,
+    tenants_dir_fd: &mut SyncWrapper<Dir>,
+    cancel: &CancellationToken,
+) -> anyhow::Result<IterationOutcome> {
+    let usage_pre = filesystem_level_usage::get(tenants_dir_fd)?.with_pressure(task_config);
+    // planned post-eviction usage
+    let mut usage_expected = usage_pre.usage.clone();
+    // achieved post-eviction usage according to internal accounting
+    let mut usage_assumed = usage_pre.usage.clone();
+    // actual usage read after batched evictions
+    let usage_post;
+
+    debug!(?usage_pre, "disk usage");
+
+    if !usage_pre.has_pressure() {
+        return Ok(IterationOutcome::NoPressure);
+    }
+
+    warn!(
+        ?usage_pre,
+        "running disk usage based eviction due to pressure"
+    );
+
+    let mut lru_candidates: Vec<(_, LocalLayerInfoForDiskUsageEviction)> = Vec::new();
+
+    // get a snapshot of the list of tenants
+    let tenants = tenant::mgr::list_tenants()
+        .await
+        .context("get list of tenants")?;
+
+    {
+        let mut tmp = Vec::new();
+        for (tenant_id, _state) in tenants {
+            let flow = extend_lru_candidates(
+                task_config,
+                tenant_id,
+                &mut lru_candidates,
+                &mut tmp,
+                cancel,
+            )
+            .await;
+
+            if let ControlFlow::Break(()) = flow {
+                return Ok(IterationOutcome::Cancelled);
+            }
+
+            assert!(tmp.is_empty(), "tmp has to be fully drained each iteration");
+        }
+    }
+
+    if cancel.is_cancelled() {
+        return Ok(IterationOutcome::Cancelled);
+    }
+
+    // phase1: select victims to relieve pressure
+    lru_candidates.sort_unstable_by_key(|(_, layer)| layer.last_activity_ts);
+    let mut batched: HashMap<_, Vec<LocalLayerInfoForDiskUsageEviction>> = HashMap::new();
+    for (timeline, layer) in lru_candidates {
+        // TODO: overshoot 1.5x to account for failing evictions
```

Comment:
TODO thread. I don't think it's controversial, just need to impl this, we didn't have energy yesterday.

Providing the 1.5x overshoot doesn't really fit naturally here because we'd have to ... `(layer.file_size() * 2) / 3` or something more correct mathematically. I think this might be best left after we have the testing support. Also the setter is quite ackward here.

Reply:
We won't have overshoot because the loop will be configured to run every 10ish seconds.